modules/gateways/twispay/lib/Twispay_Response.php

```diff
@@ -0,0 +1,387 @@
+                return TRUE;
+            break;
+
+            case self::$resultStatuses['IN_PROGRESS']:
+            case self::$resultStatuses['COMPLETE_OK']:
+                /** Add payment. */
+                addInvoicePayment($invoiceId, $response['transactionId'], $response['amount'], /*fees*/0, /*gateway*/'twispay');
+
+                /** Check if the response is for a subscription. */
+                if ('r' == $response['identifier'][0]) {
+                    self::saveSubscriptionId($invoiceId, $response);
+                }
+
+                /** Save the transaction. */
+                logTransaction(/*gatewayName*/'twispay', /*debugData*/['response' => $response, 'status' => $response['status']], __FUNCTION__ . '::' . 'Twispay IPN PROCESS: ' . Twispay_Notification::translate('TWISPAY_STATUS_SUCCESS') . $invoiceId);
+                return TRUE;
+            break;
+
+            default:
+                /** Save the transaction. */
+                logTransaction(/*gatewayName*/'twispay', /*debugData*/['response' => $response, 'status' => $response['status']], __FUNCTION__ . '::' . 'Twispay IPN PROCESS: ' . Twispay_Notification::translate('TWISPAY_WRONG_STATUS') . $response['status']);
+                return FALSE;
+            break;
+        }
+    }
+
+
+    /**
+     * Update the status of a subscription order according to the extracted server status.
+     *
+     * @param service: The recurring profile order for which to update the status.
+     * @param serverStatus: The status received from server.
+     */
+    public static function updateSubscriptionStatus($service, $serverStatus){
+        switch ($serverStatus) {
+            case $this->resultStatuses['COMPLETE_FAIL']: /* The subscription has payment failure. */
+            case $this->resultStatuses['THREE_D_PENDING']: /* The subscription has a 3D pending payment. */
+                /** Log */
+                logTransaction(/*gatewayName*/'twispay', /*debugData*/['serviceId' => $service->id, 'serverStatus' => $rserverStatus], __FUNCTION__ . '::' . ' Twispay subscription update: ' . Twispay_Notification::translate('TWISPAY_SERVER_STATUS') . $serverStatus);
+            break;
+
+            case $this->resultStatuses['COMPLETE_OK']: /* The subscription has been completed. */
+            case $this->resultStatuses['CANCEL_OK']: /* The subscription has been canceled. */
+            case $this->resultStatuses['REFUND_OK']: /* The subscription has been refunded. */
+            case $this->resultStatuses['VOID_OK']: /* The subscription has been voided. */
+            case $this->resultStatuses['CHARGE_BACK']: /* The subscription has been forced back. */
+                /** Cancel subscription. */
+                changeOrderStatus($service->order->id, 'Cancelled', /*cancelSubscription*/TRUE);

```

Comment:
I guess `$this->resultStatuses['COMPLETE_OK']` should not result in a 'Cancelled' order status.